### PR TITLE
iOS 10 (and Android) notifications

### PR DESCRIPTION
**Pull request by mrucker**
I think I got it all this time. Everything looks clean. Let me know if anything looks amiss.

### CODE REVIEW

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationCallbackScheduler.cs [hunk 1]
```diff
@@ -13,77 +13,85 @@
...
+                };
+
+                new iOSCallbackData(callbackNotification)
+                {
+                    CallbackId = meta.CallbackId,
+                    IsRepeating = meta.IsRepeating,
+                    RepeatDelay = meta.RepeatDelay,
+                    LagAllowed = meta.LagAllowed,
+                    ActivationId = SensusContext.Current.ActivationId
                 };
```
Comment (MatthewGerber) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationCallbackScheduler.cs]: Why is the above unassigned?

Comment (MatthewGerber) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationCallbackScheduler.cs]: Never mind. I see what's going on.

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs [hunk 2]
```diff
@@ -37,22 +39,6 @@ public override void IssueNotificationAsync(string message, string id, bool play
...
-                        if (notificationId == id)
-                        {
-                            CancelNotification(notification, NOTIFICATION_ID_KEY);
-
-                            // TODO:  How do we prevent this collection from growing without bound?
-                            _notifications.Remove(notification);
-                        }
-                    }
-                }
-
```
Comment (MatthewGerber) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: Where did the above cancel code go? I think we need it.

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs [hunk 3]
```diff
@@ -62,7 +48,7 @@ public override void IssueNotificationAsync(string message, string id, bool play
                         AlertBody = message,
                         TimeZone = null,  // null for UTC interpretation of FireDate
                         FireDate = DateTime.UtcNow.ToNSDate(),
-                        UserInfo = new NSDictionary(NOTIFICATION_ID_KEY, id)
```
Comment (MatthewGerber) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: Without an id embedded in the notification, we won't know which one to cancel later...?

Looks like we're conflating scheduled callbacks with notifications in this PR. Probably not a good idea. Notifications do not have delays, repeats, etc. that scheduled callbacks have. I'll have to think about this.

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationCallbackScheduler.cs (hunk 1 above)
Comment (mrucker) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationCallbackScheduler.cs]: That's actually the exact code I had in mind when I mentioned not being completely happy with this class while we were meeting today. I was just playing around with a new pattern to see how it felt. In the end I think I've decided it is more confusing than helpful.

File: Sensus.Shared/Callbacks/CallbackScheduler.cs [hunk 4]
```diff
@@ -271,6 +219,30 @@ public void UnscheduleCallback(string callbackId)
...
+            }
+
+            lock (scheduledCallback)
+            {
+                if (scheduledCallback.Running)
+                {
+                    return true;
+                }
+
+                scheduledCallback.Running = true;
```
Comment (MatthewGerber) [inline Sensus.Shared/Callbacks/CallbackScheduler.cs]: Unexpected side effect for caller.

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs (hunk 2 above)
Comment (mrucker) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: That's actually one the code changes I'm most iffy about...

I chose to remove NOTIFICATION_ID_KEY everywhere for three reasons.

1. When I did a find all on NOTIFICATION_ID_KEY I only found one "setter" which was a constant string used to identify if notifications were survey callbacks. I moved all these checks over to the new NotificationType enum.
2. We seem to use it inconsistently. Above the "id" that is passed in is the callback id (if I remember correctly). If we need the reset code maybe we should actually do a check against the CallbackId?
3. The name doesn't match the pattern of any of the other Keys (e.g. SENSUS_CALLBACK_ID_KEY). This made me think maybe it was added later after the fact without a lot of thought for how the rest of the callback code works? That may have been an incorrect assumption.

Those three aren't definitive.  I didn't find clear errors. I just found lots and lots of things that looked wrong to me. I kind of decided that removing it and fixing any bugs that might come up would be better than keeping it in.

Comment (MatthewGerber) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: It was a little convoluted before, but it worked.

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs (hunk 3 above)
Comment (mrucker) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: See my comment above. The conflation of NOTIFICATION_ID_KEY with SENSUS_CALLBACK_ID_KEY seemed to exist before our refactor? Where I felt I knew how to address I did. In other places where it appeared things didn't work as intended to me I removed. You certainly know the intent of these two id's better than me.

File: Sensus.Shared.iOS/Callbacks/iOSCallbackData.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        {
+            _data = new NSDictionary();
+        }
+
+        public iOSCallbackData(UNNotificationResponse data) : this(data?.Notification?.Request?.Content?.UserInfo)
+        { }
+
+        public iOSCallbackData(UILocalNotification local)
+        {
+            _data = local.UserInfo = new NSDictionary();
```
Comment (mrucker) [inline Sensus.Shared.iOS/Callbacks/iOSCallbackData.cs]: Ahh yes. You are right. Should probably be something like this.

`_data = local.UserInfo = local.UserInfo ?? new NSDictionary();`

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs (hunk 3 above)
Comment (MatthewGerber) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: It was working as intended previously. The conflation I see is that we're using the new callback data interface in the notifier code. These are two separate things.

File: Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs (hunk 2 above)
Comment (mrucker) [inline Sensus.Shared.iOS/Callbacks/UILocalNotifications/UILocalNotificationNotifier.cs]: Yeah I think in this case I definitely introduced a bug. I think we should simply put in CallbackId and go with that. Thoughts?

**Pull request closed without merging**